Flaky test: ThreadAffinityTest#testThreadAffinity

### TestNG Version

The latest

### Expected behavior

The test should not fail.

### Actual behavior

https://travis-ci.org/github/cbeust/testng/jobs/772083005#L1270-L1274

```
FAILURE   0.1sec, test.thread.parallelization.ThreadAffinityTest > testThreadAffinity[1]([Ljava.lang.Class;@8402514)
    java.lang.AssertionError: 
    Expected size:<2> but was:<1> in:
    <["5904", "5903"]>
        at test.thread.parallelization.ThreadAffinityTest.testThreadAffinity(ThreadAffinityTest.java:39)
```

@krmahadevan , is there a possibility you remember something regarding the test? ;)

I did try increasing the number of invocations for the test to 10000. Then I configured `LogGatheringListener` to use `Collections.synchronizedSet(Sets.newHashSet())` instead of the regular `Sets.newHashSet()`.

It caused quite a surprising result (one or two among 10000 invocations, Java 15)):

```
Expected size:<2> but was:<3> in:
<["42896", "9448", "42897"]>
java.lang.AssertionError: 
Expected size:<2> but was:<3> in:
<["42896", "9448", "42897"]>
	at test.thread.parallelization.ThreadAffinityTest.testThreadAffinity(ThreadAffinityTest.java:39)
	at jdk.internal.reflect.GeneratedMethodAccessor13.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:136)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:593)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:175)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:49)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:826)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:148)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:147)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:129)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.testng.TestRunner.privateRun(TestRunner.java:794)
	at org.testng.TestRunner.run(TestRunner.java:596)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:421)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:415)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:376)
	at org.testng.SuiteRunner.run(SuiteRunner.java:320)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1220)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1142)
	at org.testng.TestNG.runSuites(TestNG.java:1071)
	at org.testng.TestNG.run(TestNG.java:1039)
	at org.gradle.api.internal.tasks.testing.testng.TestNGTestClassProcessor.runTests(TestNGTestClassProcessor.java:141)
	at org.gradle.api.internal.tasks.testing.testng.TestNGTestClassProcessor.stop(TestNGTestClassProcessor.java:90)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:133)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

### Test case sample

`test.thread.parallelization.ThreadAffinityTest#testThreadAffinity`

@vlsi - Your understanding in terms of the thread affinity is correct. 

> Would it be better to track thread ID in the test itself rather than in the listener?

`afterInvocation()` is guaranteed to run on the same thread as the test. So it shouldnt be the cause. Moreover, we are trying to run two or more test classes together to check for the thread affinity which is what is failing.

What confuses me is this output in the test results

```
java.lang.AssertionError: 
Expected size:<4> but was:<5> in:
<[6019, 6020, 6021, 6022]>
```

How is the output saying the same was 5, when there are just 4 elements in the set ? Or am I reading this incorrectly ?

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,6 @@
 Current
+Fixed: GITHUB-2557: Flaky test: ThreadAffinityTest#testThreadAffinity (Krishnan Mahadevan)
+Fixed: GITHUB-2567: MethodHelper#CANONICAL_NAME_CACHE is never reset, so it could result in a memory leak (Krishnan Mahadevan)
 Fixed: GITHUB-2540: assertEquals(Collection) need check order (Stuart Marks & Julien Herr)
 Fixed: GITHUB-2360: Groovy 3 internal generated methods are detected as test methods (Ian Springer)
 Fixed: GITHUB-2522: TestNG 7.4.0 Can not skip test through listener (Nan Liang)

--- a/core/src/main/java/org/testng/TestRunner.java
+++ b/core/src/main/java/org/testng/TestRunner.java
@@ -1,5 +1,6 @@
 package org.testng;
 
+import java.lang.reflect.Method;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -15,6 +16,7 @@ import java.util.concurrent.PriorityBlockingQueue;
 import java.util.concurrent.BlockingQueue;
 
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
 import org.testng.collections.Lists;
 import org.testng.collections.Maps;
 import org.testng.collections.Sets;
@@ -923,6 +925,20 @@ public class TestRunner
         itl.onFinish(this);
       }
     }
+    if (!isStart) {
+      MethodHelper.clear(methods(this.getPassedConfigurations()));
+      MethodHelper.clear(methods(this.getFailedConfigurations()));
+      MethodHelper.clear(methods(this.getSkippedConfigurations()));
+      MethodHelper.clear(methods(Arrays.stream(this.getAllTestMethods())));
+    }
+  }
+
+  private static Stream<Method> methods(IResultMap resultMap) {
+    return methods(resultMap.getAllMethods().stream());
+  }
+
+  private static Stream<Method> methods(Stream<ITestNGMethod> methods) {
+    return methods.map(each -> each.getConstructorOrMethod().getMethod());
   }
 
   /////

--- a/core/src/main/java/org/testng/internal/MethodHelper.java
+++ b/core/src/main/java/org/testng/internal/MethodHelper.java
@@ -13,6 +13,7 @@ import java.util.concurrent.atomic.AtomicReference;
 import java.util.regex.Pattern;
 
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 import org.testng.IMethodInstance;
 import org.testng.ITestClass;
 import org.testng.ITestNGMethod;
@@ -454,6 +455,11 @@ public class MethodHelper {
         .findFirst()
         .map(MethodHelper::calculateMethodCanonicalName)
         .orElse(null);
+  }
+
+  public static void clear(Stream<Method> methods) {
+    methods.filter(Objects::nonNull)
+        .forEach(CANONICAL_NAME_CACHE::remove);
   }
 
   public static long calculateTimeOut(ITestNGMethod tm) {

--- a/core/src/test/java/test/thread/parallelization/ThreadAffinityTest.java
+++ b/core/src/test/java/test/thread/parallelization/ThreadAffinityTest.java
@@ -4,6 +4,7 @@ import org.assertj.core.api.Assertions;
 import org.testng.TestNG;
 import org.testng.annotations.AfterClass;
 import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
 import org.testng.internal.RuntimeBehavior;
@@ -26,6 +27,11 @@ public class ThreadAffinityTest extends SimpleBaseTest {
     System.setProperty(RuntimeBehavior.TESTNG_THREAD_AFFINITY, "true");
   }
 
+  @BeforeMethod
+  public void beforeMethod() {
+    LogGatheringListener.reset();
+  }
+
   @Test(dataProvider = "dp1")
   public void testThreadAffinity(Class<?>... classes) {
     XmlSuite xmlsuite = createXmlSuite("test_suite");
@@ -33,10 +39,8 @@ public class ThreadAffinityTest extends SimpleBaseTest {
     xmlsuite.setThreadCount(6);
     createXmlTest(xmlsuite, "Test 1", classes);
     TestNG testng = create(xmlsuite);
-    LogGatheringListener listener = new LogGatheringListener();
-    testng.addListener(listener);
     testng.run();
-    assertThat(listener.getLog()).hasSize(2);
+    assertThat(LogGatheringListener.getLog()).hasSize(2);
   }
 
   @Test(dataProvider = "dp2")
@@ -47,10 +51,8 @@ public class ThreadAffinityTest extends SimpleBaseTest {
     createXmlTest(xmlsuite, "Test_1", PriorityTestSample1.class, PriorityTestSample2.class).setParallel(mode);
     createXmlTest(xmlsuite, "Test_2", PriorityTestSample1.class,PriorityTestSample2.class).setParallel(mode);
     TestNG testng = create(xmlsuite);
-    LogGatheringListener listener = new LogGatheringListener();
-    testng.addListener(listener);
     testng.run();
-    assertThat(listener.getLog()).hasSize(size);
+    assertThat(LogGatheringListener.getLog()).hasSize(size);
   }
 
   @DataProvider(name = "dp2")

--- a/core/src/test/java/test/thread/parallelization/issue1773/LogGatheringListener.java
+++ b/core/src/test/java/test/thread/parallelization/issue1773/LogGatheringListener.java
@@ -1,23 +1,21 @@
 package test.thread.parallelization.issue1773;
 
-import org.testng.IInvokedMethod;
-import org.testng.IInvokedMethodListener;
-import org.testng.ITestResult;
-import org.testng.Reporter;
-import org.testng.collections.Sets;
-
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 
-public class LogGatheringListener implements IInvokedMethodListener {
+public class LogGatheringListener {
 
-    Set<String> log = Sets.newHashSet();
+    private static Set<Long> log = ConcurrentHashMap.newKeySet();
 
-    public Set<String> getLog() {
+    public static Set<Long> getLog() {
         return log;
     }
 
-    @Override
-    public void afterInvocation(IInvokedMethod method, ITestResult testResult) {
-        log.addAll(Reporter.getOutput(testResult));
+    public static void addLog(long threadId) {
+        log.add(threadId);
+    }
+
+    public static void reset() {
+        log = ConcurrentHashMap.newKeySet();
     }
 }

--- a/core/src/test/java/test/thread/parallelization/issue1773/MethodDependenciesSample1.java
+++ b/core/src/test/java/test/thread/parallelization/issue1773/MethodDependenciesSample1.java
@@ -16,6 +16,6 @@ public class MethodDependenciesSample1 {
   }
 
   private void log() {
-    Reporter.log(Long.toString(Thread.currentThread().getId()));
+    LogGatheringListener.addLog(Thread.currentThread().getId());
   }
 }

--- a/core/src/test/java/test/thread/parallelization/issue1773/MethodDependenciesSample2.java
+++ b/core/src/test/java/test/thread/parallelization/issue1773/MethodDependenciesSample2.java
@@ -16,6 +16,6 @@ public class MethodDependenciesSample2 {
   }
 
   private void log() {
-    Reporter.log(Long.toString(Thread.currentThread().getId()));
+    LogGatheringListener.addLog(Thread.currentThread().getId());
   }
 }

--- a/core/src/test/java/test/thread/parallelization/issue1773/PriorityTestSample1.java
+++ b/core/src/test/java/test/thread/parallelization/issue1773/PriorityTestSample1.java
@@ -1,6 +1,5 @@
 package test.thread.parallelization.issue1773;
 
-import org.testng.Reporter;
 import org.testng.annotations.Test;
 
 public class PriorityTestSample1 {
@@ -21,6 +20,6 @@ public class PriorityTestSample1 {
   }
 
   private void log(){
-    Reporter.log(Long.toString(Thread.currentThread().getId()));
+    LogGatheringListener.addLog(Thread.currentThread().getId());
   }
 }

--- a/core/src/test/java/test/thread/parallelization/issue1773/PriorityTestSample2.java
+++ b/core/src/test/java/test/thread/parallelization/issue1773/PriorityTestSample2.java
@@ -21,7 +21,7 @@ public class PriorityTestSample2 {
   }
 
   private void log(){
-    Reporter.log(Long.toString(Thread.currentThread().getId()));
+    LogGatheringListener.addLog(Thread.currentThread().getId());
   }
 
 }

